### PR TITLE
refactor and gitignore

### PR DESCRIPTION
**Pull request by Jakehp**
Well I was looking at https://github.com/ncla/LoungeDestroyer/issues/158 then got sidetracked. I made pricecheck.js a bit more functional, easier to read, and test (we'll do that...eventually...).

Make sure to test this out a bit :godmode:

### CODE REVIEW

**Comment by ncla** [comment]
`Uncaught TypeError: undefined is not a function`

File: app/pricecheck.js [hunk 1]
```diff
@@ -1,37 +1,50 @@
...
-                            return false;
-                        }
-                    });
-                    if(!itemFound) {
-                        alert(itemName + ' was not found in CSGOLounge.com database');
-                    }
-                }
-            });
+        $("#largeiteminfo_item_actions").append(CheckCSGLHtml);
+        $("#csglpricecheck").click($.ajax({url: GetCSGLSchemaUrl, type: "GET", success: onGetPriceSuccess, error: onGetPriceFailure}));
```
Comment (ncla) [inline app/pricecheck.js]: This one-liner looks confusing and moving success and error callback function to a separate place makes it even more confusing.

**Comment by Jakehp** [comment]
How do I test out my changes? Do we have the process on how to build the extension from source somewhere?

I wrote this code 'freehanded' hue

File: app/pricecheck.js [hunk 2]
```diff
@@ -1,37 +1,50 @@
...
 chrome.storage.local.get(['currencyConversionRates'], function(result) {
+
+    var GetCSGLSchemaUrl = 'http://csgolounge.com/api/schema.php';
+    var CheckCSGLHtml = '<span class="btn_small btn_grey_white_innerfade" id="csglpricecheck"><span>Check CSGOLounge.com item value</span></span>';
+
     currencies = result.currencyConversionRates || {};
-    LoungeUser.loadUserSettings(function() {
+    LoungeUser.loadUserSettings(renderAndBindGetPrice);
+
+    function renderAndBindGetPrice(){
```
Comment (ncla) [inline app/pricecheck.js]: I still don't get what's the purpose of wrapping every small thing into its own function. Please explain. pricecheck.js does only one thing and is only CS:GO specific.

File: app/pricecheck.js [hunk 3]
```diff
@@ -1,37 +1,50 @@
 var itemName = document.title.split("Listings for ")[1];
-
 var LoungeUser = new User();
+
 chrome.storage.local.get(['currencyConversionRates'], function(result) {
+
+    var GetCSGLSchemaUrl = 'http://csgolounge.com/api/schema.php';
```
Comment (ncla) [inline app/pricecheck.js]: The variable name suggests that it is a function, which it is not.

**Comment by ncla** [comment]
http://superuser.com/a/247654

**Comment by ncla** [comment]
I don't think this is easier to read, rather the opposite - it is confusing now.

@birjolaxew any input on this?

**Comment by Jakehp** [comment]
Hm, I don't see how it's confusing. Will write something up in 30 mins (omw to work) with my explanation of changes and respond to your concerns. I should have explained my reasoning in original PR.

I'll also run the original and refactor past some devs to get more input on which is clearer.

**Comment by ncla** [comment]
Sure, let us know when you got input from other devs.

File: app/pricecheck.js (hunk 3 above)
Comment (Jakehp) [inline app/pricecheck.js]: The Upper case format should indicate it's not. However; there's an even better way to distinguish the var, apparently[1] all caps snake case is the proper way to go.

E.g. CSGL_SCHEMA_ENDPOINT or CSGL_SCHEMA_URL

[1] - http://stackoverflow.com/questions/130396/are-there-constants-in-javascript
Soon - https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/const

**Comment by birjolaxew** [comment]
Gotta agree with @ncla. While I generally approve of refactoring, it made it a bit more cluttered this time around.

Before it was a pretty simple workflow:
```
add button
attach click handler to button
    load schema.php
        loop through each item
            if item name matches, display data to user
       if no match was found, display error to user
```

Now it's:
```
add button
attach click handler to button
    load schema.php
        if success:
            go to "onGetPriceSuccess"
        if error:
            go to "onGetPriceFailure"

function onGetPriceSuccess:
    if itemName exists:
        call feedbackWithValue with value of findItemValueByItemName called with values of data and itemName

function OnGetPriceFailure:
    display error to user

function feedbackWithValue:
    display relevant data to user

function findItemValueByName:
    loop through each item:
        if match, return value
    if no match, return undefined
```

I feel like there are too many proxy functions - `onGetPriceFailure` is literally a single alert statement with a static string in it, `findItemValueByName` is a simple `forEach` loop, and `onGetPriceScucess`, `feedbackWithValue` and `findItemValueByName` could easily be combined into one function without hurting readability (rather helping it)

That being said, there are some things in here we need for the next update: handling errors when loading `schema.php`, most notably.

**Comment by Jakehp** [comment]
Interesting, feedback I received from other's was that what I suggested is more clear. Would you guys agree that there is less coupling in the changes I suggested?

**Comment by birjolaxew** [comment]
I'd prefer kind of a mix. We've got no documentation in the form of function names/comments in ours, you refactor it into tiny pieces.

If you group all success-related logic into one function (since you currently have one that basically is an `if` statement and another that's basically a `forEach` loop) and all error-related logic into another, that'd be a lot clearer without being as fragmented.

**Comment by Jakehp** [comment]
Testability (debatable)
Although everything is in an anonymous function, how would you ever go about testing this?
```js
$("#csglpricecheck").click(function() {
  var itemFound = false;
  $.ajax({
      url: "http://csgolounge.com/api/schema.php",
      type: "GET",
      success: function(data){
          $.each(data, function(i, v) {
              if(v.name == itemName) {
                  var worth = parseFloat(v.worth).toFixed(2);
                  itemFound = true;
                  if(worth > 0) {
                      alert(itemName + ' is worth ' + convertPrice(worth, true) + ' on CSGOLounge.com');
                  } else {
                      alert(itemName + ' is not available for betting on CSGOLounge.com');
                  }
                  return false;
              }
          });
          if(!itemFound) {
              alert(itemName + ' was not found in CSGOLounge.com database');
          }
      }
  });
});
```
With the changes in place you could easily test your lookup, or feedback func.

Maintainability
Want to quickly change the way messages/feedback is given? In the previous way, you would add more code into this ajax on success block.
Need to add complexity to your lookup? In the previous way, you would add more code into this ajax on success block.
Basically the idea I went for is prevention of some gross deeply nested js block.

Optimization
From what I've heard from a friend, chrome optimizes functions if not anonymous. Would need to look up sources.

File: app/pricecheck.js (hunk 1 above)
Comment (Jakehp) [inline app/pricecheck.js]: @ncla
```js
$("#csglpricecheck").click($.ajax({url: GetCSGLSchemaUrl, type: "GET", success: onGetPriceSuccess, error: onGetPriceFailure}));
```
'''
This one-liner looks confusing and moving success and error callback function to a separate place makes it even more confusing.
'''
Using predefined functions for onSuccess/Fail is not an abnormal thing and I don't think it really causes confusion.

**Comment by Jakehp** [comment]
@birjolaxew 
"onGetPriceFailure is literally a single alert statement with a static string in it"
Well besides considering *nothing* was being done on error; do you only consider what currently exists to make a decision whether that functionality should be on it's own? It's also about maintainability and extensibility for the future.

**Comment by birjolaxew** [comment]
So do this:

```javascript
$("#csglpricecheck").click(function() {
  var itemFound = false;
  $.ajax({
      url: "http://csgolounge.com/api/schema.php",
      type: "GET",
      success: lookupItem,
      error: function(){alert("Couldn't connect to CSGOLounge")}
  });
});

function lookupItem(itemList) {
  // loop through Lounge's item list
  for (var item in itemList) {
    if (item.name === itemName) {
        // if item match, display data to user
        var worth = parseFloat(v.worth).toFixed(2);
        if(worth > 0) {
          alert(itemName + ' is worth ' + convertPrice(worth, true) + ' on CSGOLounge.com');
          return;
        }
    }
  }
  
  // if no match
  alert(itemName + ' is not available on CSGOLounge.com');
}
```

Instead of this (exaggerated so you get the point):

```javascript
$("#csglpricecheck").click(get(GetCSGLSchemaUrl, onGetPriceSuccess, onGetPriceFailure));

function onGetPriceSuccess(itemSet){
  if(checkIfVarIsTruthy(itemName)){
    feedbackWithItemValue(findItemValueByItemName(itemSet, itemName));
  } else {
    alertString('No item name set, unable to lookup value.');
  }
}

function onGetPriceFailure(){
  alertString('CSGOLounge.com data is unavailable.');
}

function feedbackWithItemValue(itemValue){
  if (checkIfVarIsTruthy(itemValue) && itemValue > 0){
    alertStr(itemName + ' is worth ' + convertPrice(itemValue, true) + ' on CSGOLounge.com');
  } else if (checkIfVarIsTruthy(itemValue) && itemValue <= 0){
    alertStr(itemName + ' is not available for betting on CSGOLounge.com');
  } else {
    alertStr(itemName + ' was not found in CSGOLounge.com schema');
  }
}

function findItemValueByItemName(itemSet, itemName){
  if(checkIfVarIsTruthy(checkitemSet)){ return undefined; }
    loopObject(itemSet, function(item){
      if(checkIfVarIsTruthy(item) && checkIfVarIsTruthy(item.name) && item.name == itemName && checkIfVarIsTruthy(item.worth)) {
        return parseNumber(item.worth, 2);
      }
    });
    return undefined;
}

function checkIfVarIsTruthy(variable) {
  return !!variable;
}

function alertString(str) {
  alert(str);
}

function parseNumber(num, digits) {
  return parseFloat(num).toFixed(2);
}

function loopObject(obj, callback) {
  obj.forEach(callback);
}

function get(url, successCallback, errorCallback, options) {
  options.url = url;
  options.success = successCallback;
  options.error = errorCallback;
  $.ajax(options);
}
```

**Comment by Jakehp** [comment]
I'd go with the second, *if*, there was structure in the app to support more code reuse. Obviously not just defining a list of functions in the global context or whatever.

Except for loopObject, lol.

**Comment by birjolaxew** [comment]
Where on earth would we reuse:
```javascript
function onGetPriceFailure(){
  alert('CSGOLounge.com data is unavailable.');
}
```
or
```javascript
function onGetPriceSuccess(itemSet){
  if(itemName){
    feedbackWithItemValue(findItemValueByItemName(itemSet, itemName));
  } else {
    alert('No item name set, unable to lookup value.');
  }
}
```

**Comment by Jakehp** [comment]
For those - https://github.com/ncla/LoungeDestroyer/pull/161#issuecomment-71684571

My last comment was more directed at your exaggeration functions that you added.

**Comment by Jakehp** [comment]
Throwing a single if statement into a function for itself does not improve readability.
```js
function onGetPriceSuccess(itemSet){
  if(itemName){
    feedbackWithItemValue(findItemValueByItemName(itemSet, itemName));
  } else {
    alert('No item name set, unable to lookup value.');
  }
}
```
Nope it doesn't, it prevents an undefined error until we fix the itemName bug. Once that is fixed i'd probably put
```
feedbackWithItemValue(findItemValueByItemName(itemSet, itemName));
```
into the ajax block.

**Comment by birjolaxew** [comment]
The `if` statement prevents the undefined error. Moving it to a seperate function has nothing to do with that.

Heck, the check should happen before the user clicks the button - if we fail at parsing the item name, we shouldn't display the button.

My previously suggested function does exactly the same as yours, except without moving basic JS functionality into seperate functions, and it is (arguably) more readable (not to mention short).

**Comment by Jakehp** [comment]
"if we fail at parsing the item name, we shouldn't display the button." That sounds good.

**Comment by ncla** [comment]
So I see you have closed all the issues and PRs, just because children in a sandbox are bragging about who has the best sandcastle and can't understand the fact that the sandcastle does the same thing as it did before. Refactoring here was almost pointless and wasted effort.

pricecheck.js serves only one purpose and nothing more. It might as well be tied with app/items.js (with additional method checkCSGLValue on items.js then).

The only valid arguments you brought up that are legitimate in this case were testability and extensibility, neither of which we are sure we really want yet (that's why the issue regarding unit testing is still up in the air). Not sure about extensibility since I do not know where our application could be really be extended apart from creating user-scripts is based on our application files.

Too bad to easy you go because of different opinions and disagreement in this pull request.

**Comment by Jakehp** [comment]
In this sandbox we try to be respectful of their others and not sarcastic. After talking with the other developer, I didn't appreciate his attitude and I no longer wish to be involved. Nothing wrong with that.

**Comment by birjolaxew** [comment]
If you feel like I attacked you personally, I'm sorry. That obviously wasn't the intention.

Most of what you quoted was me trying to explain, for the 4-5th time, why I disliked parts of your refactoring. You moved basic JS functionality into functions for themselves, which only harms readability.
Since you didn't understand my point after I provided pseudo-code, direct instructions *and* my own refactoring which ticked all your wishes and didn't fragmentize the code as much, I tried using direct examples from your code to show what I disliked.

Constructive criticism is the basis of open-source code. While I agree that I might have stopped with the nice tone, you taking it as a personal attack and deleting all your contributions seems kind of... childish.

**Comment by Jakehp** [comment]
"If you feel like I attacked you personally, I'm sorry."
Much appreciated.
"While I agree that I might have stopped with the nice tone, *you taking it* as a personal attack"
Oh wait..

"deleting all your contributions seems kind of... childish."
What are you talking about? My contributions are merged, and anything unmerged you can find in their respective place. E.g. https://github.com/ncla/LoungeDestroyer/pull/159 - https://github.com/ncla/LoungeDestroyer/commit/bbcca50a64539b19fd665d424adcc035037a197b - https://github.com/ncla/LoungeDestroyer/commit/5668b49467abee6fbfde4506cf34cb64a3acfc74
I closed the issues because I'm no longer involved.
